### PR TITLE
New stream builder

### PR DESCRIPTION
**Pull request by semenodm**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -113,6 +113,7 @@ lazy val `fs2-aws-kinesis` = (project in file("fs2-aws-kinesis"))
     Dependencies.Fs2Core,
     Dependencies.KinesisClient,
     Dependencies.KinesisProducer,
+    Dependencies.newTypes,
```
Comment (barryoneill) [inline build.sbt:116]: are we moving away from refined types?

Comment (semenodm) [reply build.sbt:116]: i'm not sure if they are compatible with scala 3

File: fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/KinesisStreamBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import software.amazon.awssdk.services.cloudwatch.CloudWatchAsyncClient
+import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient
+import software.amazon.awssdk.services.kinesis.KinesisAsyncClient
+import software.amazon.kinesis.common.ConfigsBuilder
+import software.amazon.kinesis.coordinator.Scheduler
+import software.amazon.kinesis.processor.{MultiStreamTracker, SingleStreamTracker, StreamTracker}
+
+abstract class KinesisStreamBuilder[F[_]] {
+
+  trait InitialPhase {
```
Comment (barryoneill) [inline fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/KinesisStreamBuilder.scala:16]: this feels like a pretty restrictive builder, in that you have to call each phase, in a particular order, before `build` is available?

Comment (semenodm) [reply fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/KinesisStreamBuilder.scala:16]: correct, it has one branching, depending on what type of stream tracker you want, single stream or multistream

Comment (barryoneill) [reply fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/KinesisStreamBuilder.scala:16]: Okay, but if you're forcing the caller to provide all the required `with` calls up to that branch, i.e.

``` scala
        .withAppName(appName)
        .withKinesisClient(k)
        .withDynamoDBClient(DynamoDbInterpreter[F].DynamoDbAsyncClientResource(dac))
        .withCloudWatchClient(CloudwatchInterpreter[F].CloudWatchAsyncClientResource(cac))

```

then that could all be provided in one function call saving a lot of boilerplate.  Or, put that behind a convenience function on the KinesisStreamBuilder object, e.g.

```scala
KinesisStreamBuilder.default[F](appName, kClient, dbClient, cwClient)
```

Comment (barryoneill) [reply fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/KinesisStreamBuilder.scala:16]: Unless the plan is to offer alternatives in the builder (I don't see any) for those `with` calls

Comment (semenodm) [reply fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/KinesisStreamBuilder.scala:16]: also, we can do what you suggest by adding fast forward phase where all 4 params are passed and next phase would be stream tracker or scheduler, or even final phase where everything created by default (maybe need explicitly provide stream tracker as single stream and stream name to track )

Comment (barryoneill) [reply fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/KinesisStreamBuilder.scala:16]: Not particularly - go for it

**Review by barryoneill — APPROVED** [review]

**Pull request merged**